Iterative solver --- PCG implementation

nychiang opened the pull request:
Implement the PCG framework in HiOp

cnpetra commented:
is this ready for review?

nychiang commented:
> is this ready for review?
yes

cnpetra commented on src/LinAlg/hiopKrylovSolver.cpp:
```diff
@@ -0,0 +1,351 @@
+  // Starting procedure
+  //////////////////////////////////////////////////////////////////
+  // starting point
+  if(x0_) {
```
I would allocate `x0_` as soon as possible and then, if possible, reuse it for `xk_` to save on one vector

cnpetra commented on src/LinAlg/hiopKrylovSolver.cpp:
```diff
@@ -0,0 +1,351 @@
+
+  hiopPCGSolver::~hiopPCGSolver()
+  {
+    if(xk_) {
```
remove `if`, delete works fine for null pointers

cnpetra commented on src/LinAlg/hiopKrylovSolver.cpp:
```diff
@@ -0,0 +1,351 @@
+  
+  hiopKrylovSolver::~hiopKrylovSolver()
+  {
+    if(x0_) {
```
remove if

cnpetra commented on src/LinAlg/hiopKrylovSolver.hpp:
```diff
@@ -0,0 +1,151 @@
+  virtual bool solve(hiopVector& x) = 0;
+
+  /** Set the initial guess for the Krylov solver */
+  virtual bool set_x0(hiopVector& x0);
```
`const hiopVector& x`

maybe add a setter that set the initial point uniformly equal to a constant?

cnpetra commented on src/LinAlg/hiopLinearOperator.hpp:
```diff
@@ -0,0 +1,107 @@
+class hiopLinearOperator
+{
+public:
+  hiopLinearOperator(hiopMatrix* mat);
```
I would remove this constructor and `mMat_` as linear operators are not necessarily linked to matrices... a default constructor should suffice

cnpetra commented on src/LinAlg/hiopLinearOperator.hpp:
```diff
@@ -0,0 +1,107 @@
+  virtual bool times_vec(hiopVector& y, hiopVector& x) = 0;
+
+  /** y = Mat' * x */
+  virtual bool trans_times_vec(hiopVector& y, hiopVector& x) {return false;};
```
maybe call `times_vec` (assume linear operator is symmetric)? or just leave it pure virtual?

cnpetra commented on src/LinAlg/hiopLinearOperator.hpp:
```diff
@@ -0,0 +1,107 @@
+  virtual ~hiopLinearOperator() {};
+
+  /** y = Mat * x */
+  virtual bool times_vec(hiopVector& y, hiopVector& x) = 0;
```
shouldn't `x` be `const`?

nychiang commented:
@cnpetra I have addressed all the issue you mentioned, and renamed some files.
Please have a look of the implementation of `set_x(const double xval)`
